src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs

```diff
@@ -77,10 +81,11 @@ internal ComObject(IIUnknownInterfaceDetailsStrategy interfaceDetailsStrategy, I
         /// </remarks>
         public void FinalRelease()
         {
-            if (UniqueInstance)
+            if (UniqueInstance && Interlocked.CompareExchange(ref _released, 1, 0) == 0)
             {
                 CacheStrategy.Clear(IUnknownStrategy);
                 IUnknownStrategy.Release(_instancePointer);
```

Comment:
Is it possible Clear / Release could throw an exception, and if they did, what would happen when the finalizer runs and calls them again?

Reply:
Since both `CacheStrategy` and `IUnknownStrategy` can be user-defined types, they could throw exceptions. If `CacheStrategy.Clear` or `IUnknownStrategy.Release` throws an exception, then the call to `CacheStrategy.Clear` should be okay to try again (it should be possible to clear the cache multiple times). If `IUnknownStrategy.Release` throws an exception, then it shouldn't have released the COM object pointed to by `_instancePointer`, so calling `Release` again should be fine.

Reply:
> then it shouldn't have released the COM object pointed to by _instancePointer

What guarantees that? Seems like in the face of an exception we're trading off corruption vs a leak... I'd be inclined to say the leak is better. Maybe we should suppress finalization before the call instead?

Reply:
I'll do that.

Reply:
Might also be worth adding a test for the exceptional case. I would be most thrilled if we could catch corruption issues here earlier using GCStress rather than debugging user DMPs.